### PR TITLE
Tabular: Added initial RAPIDS KNN model

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

- Added initial RAPIDS KNN model
- On a p3.2xlarge, shows **1000x+** faster inference speed with zero accuracy drop compared to sklearn!!

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Innixma** [comment]
Install instructions:

```
conda create -n rapids-0.18 -c rapidsai -c nvidia -c conda-forge -c defaults rapids-blazing=0.18 python=3.7 cudatoolkit=10.1 -y
conda activate rapids-0.18
git clone -b tabular_rapids_rf https://github.com/awslabs/autogluon
cd autogluon
python3 -m pip install -e core/
python3 -m pip install -e features/
python3 -m pip install -e tabular/[all]
cd ..
```

Test code:

```

from autogluon.tabular import TabularPredictor, TabularDataset

# SAMPLE = 5000
SAMPLE = 100000
# SAMPLE = None
LABEL = 'Cover_Type'


if __name__ == '__main__':
    path_prefix = 'https://autogluon.s3.amazonaws.com/datasets/CoverTypeMulticlassClassification/'

    path_train = path_prefix + 'train_data.csv'
    path_test = path_prefix + 'test_data.csv'

    train_data = TabularDataset(path_train)
    test_data = TabularDataset(path_test)

    if SAMPLE is not None:
        train_data = train_data.sample(n=SAMPLE, random_state=0)

    # from autogluon.tabular.models.rf.rf_rapids_model import RFRapidsModel
    from autogluon.tabular.models.knn.knn_rapids_model import KNNRapidsModel

    predictor = TabularPredictor(label=LABEL, eval_metric='log_loss').fit(
        train_data=train_data,
        hyperparameters={
            # 'RF': {},
            # RFRapidsModel: {},
            'KNN': {},
            KNNRapidsModel: {},
        },
    )

    leaderboard = predictor.leaderboard(test_data)

```

Output:

```
                 model  score_test  score_val  pred_time_test  pred_time_val  fit_time  pred_time_test_marginal  pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  fit_order
0       KNNRapidsModel   -0.307182   -0.31517        0.474391       0.430484  3.345487                 0.474391                0.430484           3.345487            1       True          2
1  WeightedEnsemble_L2   -0.307186   -0.31517      503.129004      10.521359  0.657844                 0.021013                0.001806           0.585180            2       True          3
2           KNeighbors   -0.307186   -0.31517      503.107990      10.519553  0.072664               503.107990               10.519553           0.072664            1       True          1

```

**Comment by szha** [comment]
Job PR-997-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-997/3/index.html

File: tabular/src/autogluon/tabular/models/knn/knn_rapids_model.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    NOTE: This code is experimental, it is recommend to not use this unless you are a developer.
+    This was tested on rapids-0.18 via:
+
+    conda create -n rapids-0.18 -c rapidsai -c nvidia -c conda-forge -c defaults rapids-blazing=0.18 python=3.7 cudatoolkit=10.1 -y
+    conda activate rapids-0.18
+    pip install --pre autogluon.tabular[all]
+    """
+    def _get_model_type(self):
+        from cuml.neighbors import KNeighborsClassifier, KNeighborsRegressor
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/knn/knn_rapids_model.py:26]: should be lazy import?

**Comment by jwmueller** [comment]
Should the get default models function automatically choose between rapids and sklearn KNN based on num_gpus argument?

File: tabular/src/autogluon/tabular/models/knn/knn_rapids_model.py (hunk 1 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/knn/knn_rapids_model.py:26]: This is a lazy import right?

**Comment by Innixma** [comment]
> Should the get default models function automatically choose between rapids and sklearn KNN based on num_gpus argument?

This will likely be the case once we are more confident in their stability (plus we need a check for rapids install). I've documented the models as experimental, so we can change them / merge them with other models as we see fit.

File: tabular/src/autogluon/tabular/models/knn/knn_rapids_model.py (hunk 1 above)
Comment (jwmueller) [reply tabular/src/autogluon/tabular/models/knn/knn_rapids_model.py:26]: Right butI meant as an explicit try_import with an informative error message as we have for other lazy imports.

**Review by jwmueller — APPROVED** [review]

File: tabular/src/autogluon/tabular/models/knn/knn_rapids_model.py (hunk 1 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/knn/knn_rapids_model.py:29]: added

**Comment by szha** [comment]
Job PR-997-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-997/4/index.html

**Pull request merged**